### PR TITLE
Add a rerun last task command

### LINKED ISSUE #25310: Re-run Last Task
- VSCode Version: 1.11.x
- OS Version: all

There does not appear to be a way to re-run the previously executed task.  You can either run "the build" using `ctrl`+`shift`+`b`, or you must explicitly select a task to run.  In Atom and Eclipse, you can hit `F9` to re-run your last build.  This is very handy, and easily what I want to do 80%-90% of the time, just re-run my last build/task command.  However, vscode does not appear to allow you to map any hot-key to the last build/task command.  So if I want to run task `foo`, I would, effectively, have to type `task foo` into the command pallet every time. Things would go a lot faster if I could just type `task foo` into the command pallet just once, then hit a hot-key to repeat the last task.

### PR DESCRIPTION
**Pull request by alexr00**
Fixes #25310

### CODE REVIEW

File: src/vs/workbench/parts/tasks/common/taskSystem.ts [hunk 1]
```diff
@@ -85,6 +83,7 @@ export const enum TaskExecuteKind {
 export interface ITaskExecuteResult {
 	kind: TaskExecuteKind;
 	promise: TPromise<ITaskSummary>;
+	task: Task;
```
Comment (dbaeumer) [inline src/vs/workbench/parts/tasks/common/taskSystem.ts]: Can you quickly explain why we need this?

**Review by dbaeumer — APPROVED** [review]
LGTM. Only one questions.

With these changes you will own the terminal task system :-)

**Review by dbaeumer — CHANGES_REQUESTED** [review]
Actually there is on thing we should think about: would it make sense to let users control what rerunning means in a per task basis. I could for example imagine that rerunning some tasks should reevaluate the vars and rerunning some other should not. So would be cool if I can control this with a property on the task itself. The question will then be what the default is. I would actually argue to should rerun and reevaluate.

**Comment by alexr00** [comment]
As a task property, I agree that the default should be to rerun and reevaluate. Having the task revaluate seems like a logical step from Run to Rerun, while having it as a setting gives the user more power to use the already evaluated variables. I'll make this change.

File: src/vs/workbench/parts/tasks/common/taskSystem.ts (hunk 1 above)
Comment (alexr00) [reply src/vs/workbench/parts/tasks/common/taskSystem.ts]: In `handleExecuteResult` in task.contribution.ts we need the Task that was executed. For tasks that were run using rerun, task.contribution.ts has no knowledge of the task. Putting the task in the execute result solves this.

**Comment by dbaeumer** [comment]
Would it make sense to introduce a `RunOption` and have a property in there and move the isBackground there as well. I am indifferent as well but top level things on the task might get messy at some time.

**Review by dbaeumer — APPROVED** [review]

**Comment by hendrics** [comment]
@alexr00 this is a long awaited feature! Thank you for this! 
 
One quick question though - there are no tests for this feature except adding a custom section, or does it do more?

**Comment by alexr00** [comment]
I didn't write any new tests for this command, that's correct. You do not need to add anything to your tasks.json to try it out with the default rerun behavior. The default behavior is that _all_ task related variables will be re-evaluated. If you want to make your task run faster and you always want the same variables as the previous run, you can set the `runOptions` in your tasks.json to have `"rerunBehavior":"useEvaluated"`

**Comment by hendrics** [comment]
I am looking forward to trying it when it’s released! I am just curious it was accepted with no automated tests. I am wondering if it is a newer tendency and automated tests no longer considered a best practice?
Nevertheless thank you for implementing it!  I am just concerned this or other functionality might regress. And I use vscode everyday :-)!

**Comment by lonniev** [comment]
By the way, HOW is this new feature executed? I don't (think) I see it in the Ctrl-Shift-A selection set.

**Pull request merged**